From neilb@suse.de  Wed Feb 17 14:37:21 2010
From: Neil Brown <neilb@suse.de>
Date: Wed, 10 Feb 2010 12:09:33 +1100
Subject: sysfs: differentiate  between locking links and non-links
To: Eric W Biederman <ebiederm@xmission.com>
Cc: Tejun Heo <tj@kernel.org>, Greg Kroah-Hartman <gregkh@suse.de>
Message-ID: <19314.1869.847327.15190@notabene.brown>

sysfs: differentiate  between locking links and non-links for sysfs

symlinks and non-symlink is sysfs are very different.
A symlink can never be locked (active) while an attribute
modification routine is running.  So removing symlink from an
attribute 'store' routine should be permitted without any lockdep
warnings.

So split the lockdep context for 's_active' in two, one for symlinks
and other for everything else.

Signed-off-by: NeilBrown <neilb@suse.de>
Cc: Eric Biederman <ebiederm@xmission.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 fs/sysfs/dir.c   |    5 ++++-
 fs/sysfs/sysfs.h |    6 +++---
 2 files changed, 7 insertions(+), 4 deletions(-)

--- a/fs/sysfs/dir.c
+++ b/fs/sysfs/dir.c
@@ -354,7 +354,10 @@ struct sysfs_dirent *sysfs_new_dirent(co
 
 	atomic_set(&sd->s_count, 1);
 	atomic_set(&sd->s_active, 0);
-	sysfs_dirent_init_lockdep(sd);
+	if (type & SYSFS_KOBJ_LINK)
+		sysfs_dirent_init_lockdep(sd, "link");
+	else
+		sysfs_dirent_init_lockdep(sd, "non_link");
 
 	sd->s_name = name;
 	sd->s_mode = mode;
--- a/fs/sysfs/sysfs.h
+++ b/fs/sysfs/sysfs.h
@@ -89,14 +89,14 @@ static inline unsigned int sysfs_type(st
 }
 
 #ifdef CONFIG_DEBUG_LOCK_ALLOC
-#define sysfs_dirent_init_lockdep(sd)				\
+#define sysfs_dirent_init_lockdep(sd, type)			\
 do {								\
 	static struct lock_class_key __key;			\
 								\
-	lockdep_init_map(&sd->dep_map, "s_active", &__key, 0);	\
+	lockdep_init_map(&sd->dep_map, "s_active_" type, &__key, 0);	\
 } while(0)
 #else
-#define sysfs_dirent_init_lockdep(sd) do {} while(0)
+#define sysfs_dirent_init_lockdep(sd, type) do {} while(0)
 #endif
 
 /*
